### PR TITLE
Add user roles for organizations/gangs/sections

### PR DESCRIPTION
**Pull request by robines**
This PR introduces a new [custom Auth backend](https://docs.djangoproject.com/en/5.0/topics/auth/customizing/). It adds a new Role model, which basically works as a grouping of permissions. It also adds a user's connection to these Roles. This must be done on one of the following hierarchical levels: Organization, Gang, Gang Section. This is also the order the new auth backend checks for permissions.

The auth backend relies on the `has_perm` method from Django, and for our backend, the `obj` argument is required. This means the new backend only checks permissions against specific provided objects. 

For each hierarchy level, it resolves the organization, gang or section, and returns True if the user has the required permissions on that level.

## Organization/Gang/Section resolvers

The way the backend knows what organization an object belongs to is through the new `resolve_org`/`resolve_gang`/`resolve_section` methods. The purpose of these methods for an object is to return the org/gang/section which the object belongs to. For instance, a Recruitment is owned by an organization, therefore `resolve_org` returns that organization.

Not all objects have a single owner, if any at all. In that case, they simply don't implement the resolvers.

The resolvers currently have a single argument: `return_id`. The purpose of this argument is to avoid having to unnecessarily fetch a whole instance from the database, when we only need the ID. All models which implement the resolvers must respect this argument if possible.

The great thing about the introduction of these resolvers is that they may be used everywhere, not just in the new auth backend! It greatly simplifies code, and makes it a breeze to get the org/gang/section owner of a specific object.

## Example

For example's sake, say we have a "Recruitment Admin" role. It contains all relevant recruitment permissions. This role may be given to users on either the org/gang/section level (or multiple levels). If the user has the role on the Organization level, they receive the permissions for the selected organization, as well as for all the organization's gangs and roles. This means they can manage every recruitment object which belongs to the organization.

Similarly, if the user has the role on the Gang level, they receive the permissions for the selected gang, as well as for all the gang's sections. They can then manage all recruitment objects which belongs to the gang.

### CODE REVIEW

File: backend/samfundet/models/general.py [hunk 1]
```diff
@@ -132,6 +133,8 @@ class User(AbstractUser):
         on_delete=models.PROTECT,
     )
 
+    roles = models.ManyToManyField(Role, blank=True, related_name='users')
+
```
Comment (magsyg) [inline backend/samfundet/models/general.py:137]: Funker dette med kontakt?

Antar at den skal ha en rekke forskjellige mulige modeller gjennom dette
Skal den da muligens ha et through felt på UserRoleBase?

File: backend/samfundet/views.py [hunk 2]
```diff
@@ -973,12 +974,22 @@ class InterviewRoomView(ModelViewSet):
...
+    # noinspection PyMethodOverriding
     def list(self, request: Request) -> Response:
         recruitment = request.query_params.get('recruitment')
         if not recruitment:
             return Response({'error': 'A recruitment parameter is required'}, status=status.HTTP_400_BAD_REQUEST)
 
-        filtered_rooms = InterviewRoom.objects.filter(recruitment__id=recruitment)
+        filtered_rooms = [
+            room for room in InterviewRoom.objects.filter(recruitment__id=recruitment) if request.user.has_perm(SAMFUNDET_VIEW_INTERVIEWROOM, room)
+        ]
```
Comment (magsyg) [inline backend/samfundet/views.py:992]: Skal dette gjøres her?
Eller er dette for å teste?

**Review by magsyg — COMMENTED** [review]
Ser ok ut
Men sett opp tester for hver av typene
Typ hva om man har/ikke kan gjøre noe basert på org/gang/generelt

File: backend/samfundet/models/general.py (hunk 1 above)
Comment (robines) [reply backend/samfundet/models/general.py]: Denne var gjenglemt fra testing, skulle ikke være der. Om en bruker skal få tildelt en Role, må det gjøres gjennom enten UserOrgRole, UserGangRole eller UserGangSectionRole.

File: backend/samfundet/views.py (hunk 2 above)
Comment (robines) [reply backend/samfundet/views.py:992]: Tenkte det var greit å inkludere flere eksempler på hvordan systemet kan brukes, men kan fjerne om du vil? Tenkte at bruken av det nye systemet var minimal nok til å inkludere i samme PR

File: backend/root/settings/base.py [hunk 3]
```diff
@@ -87,6 +87,7 @@
     'corsheaders',
     'root',  # Register to enable management.commands.
     'samfundet',
+    'samfundet.roleauth',
```
Comment (emilte) [inline backend/root/settings/base.py:90]: Trenger det være egen app?

Comment (robines) [reply backend/root/settings/base.py]: Nei må ikke det. Har flyttet til `samfundet` nå.

File: backend/samfundet/tests/test_roles.py
Comment (magsyg) [inline backend/samfundet/tests/test_roles.py:1]: Legg til også hierarkise tester
veldig viktig at disse funker som de skal
-  Du har Org rettigheter, så kan du gjøre ting med sections og gang? Og så tester nedover her
- Du har Section rettigheter, så du kan ikke gjøre noe over, og så oppover

Også gjennomfør disse testene på noen views, men dette kan nok være i en annen PR hvor vi innfører dette per objekt nivå

File: backend/samfundet/views.py (hunk 2 above)
Comment (robines) [reply backend/samfundet/views.py]: @magsyg Har flyttet til ny PR #1319

Lukker denne tråden

File: backend/samfundet/tests/test_roles.py
Comment (robines) [reply backend/samfundet/tests/test_roles.py:1]: @magsyg Har lagt til flere tester som tester hierarkien

**Review by Snorre98 — APPROVED** [review]
lgtm

**Review by Mathias-a — COMMENTED** [review]
Har ikke slettet eksisterende funksjonalitet?

**Comment by robines** [comment]
@Mathias-a Føler det faller litt utenfor scopet til denne PRen, siden det kreves ikke for å få dette rollesystemet til å funke. Så har laget en ny issue på det #1342

**Review by magsyg — APPROVED** [review]
Looks good

**Pull request merged**